### PR TITLE
libion.so support.

### PR DESCRIPTION
**Pull request by ronlieb**
SDM845 currently supports these APIs, as well as the old IOCTL approach.
Older platforms from Qualcomm only support IOCTL approach.
Newer platforms from QUalcomm after SDM845 will only support libion.so

This patch will first look for libion.so and use it if found, otherwise fall
back to the old IOCTL approach.

### CODE REVIEW

**Comment by ronlieb** [comment]
I have omitted the rebuilt libraries, for the time being.
Once we start to converge i will build them and push up the libs.
Also, i would like to know which version Google is currently using for the Hexagon SDK, Hexagon., Tools, and Android tools.

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 1]
```diff
@@ -43,7 +46,28 @@ struct ion_handle_data {
 #define ION_IOC_FREE _IOWR('I', 1, ion_handle_data)
 #define ION_IOC_MAP _IOWR('I', 2, ion_fd_data)
 
-ion_user_handle_t ion_alloc(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags) {
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/host_malloc.cpp]: I think that all of the old changes to ion_* should be reverted, and keep all of the branching to the new API in halide_hexagon_host_*. Right now, it's kind of done in both places, but not completely in either place, which is confusing/complicated.

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 2]
```diff
@@ -43,7 +46,28 @@ struct ion_handle_data {
...
+extern "C" {
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/host_malloc.cpp]: Only the weak functions here should be extern "C", the rest should be in an anonymous namespace (internal linkage).

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 3]
```diff
@@ -69,6 +97,8 @@ int ion_map(int ion_fd, ion_user_handle_t handle) {
 }
 
 int ion_free(int ion_fd, ion_user_handle_t ion_handle) {
+    if (use_libion) return 0;
```
[the remark this answers is not in the document]
Comment (ronlieb) [reply src/runtime/hexagon_remote/host_malloc.cpp]: We have an application porting note, (i will send it to you) that has this comment:
ION_IOC_FREE
[Deprecated]
Call close on the dma_buf fd instead.

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 4]
```diff
@@ -137,19 +184,29 @@ void *halide_hexagon_host_malloc(size_t size) {
         }
     }
 
-    ion_user_handle_t handle = ion_alloc(ion_fd, size, alignment, 1 << heap_id, ion_flags);
-    if (handle < 0) {
-        __android_log_print(ANDROID_LOG_ERROR, "halide", "ion_alloc(%d, %d, %d, %d, %d) failed",
+    int buf_fd = 0;
+    ion_user_handle_t handle = 0;
+    if (use_libion) {
+        buf_fd  = ion_alloc(ion_fd, size, alignment, 1 << heap_id, ion_flags);
```
[the remark this answers is not in the document]
Comment (ronlieb) [reply src/runtime/hexagon_remote/host_malloc.cpp]: sounds good to me

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 5]
```diff
@@ -43,7 +46,28 @@ struct ion_handle_data {
...
+     __attribute__((weak)) extern int ion_alloc_fd(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags, int* map_fd);
+     typedef int (*rem_ion_alloc_fd_fn)(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags, int* map_fd);
+     rem_ion_alloc_fd_fn ion_alloc_fd_fn = NULL;
+}
+
+int ion_alloc(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags) {
+
+    if (use_libion) {
+        int map_fd =0;
+        if (ion_alloc_fd_fn(ion_fd, len, 0, heap_id_mask, flags, &map_fd)) {
```
[the remark this answers is not in the document]
Comment (ronlieb) [reply src/runtime/hexagon_remote/host_malloc.cpp]: yes should be align

File: src/runtime/hexagon_remote/host_malloc.cpp (hunk 1 above)
Comment (ronlieb) [reply src/runtime/hexagon_remote/host_malloc.cpp]: in progress, thx for suggestion

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 6]
```diff
@@ -43,7 +46,21 @@ struct ion_handle_data {
 #define ION_IOC_FREE _IOWR('I', 1, ion_handle_data)
 #define ION_IOC_MAP _IOWR('I', 2, ion_fd_data)
 
-ion_user_handle_t ion_alloc(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags) {
+// For libion.so usage
+extern "C" {
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/host_malloc.cpp]: Actually, looking more closely, I think these are unused and should be removed. You are using the function pointers below only (which makes sense).

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 7]
```diff
@@ -43,7 +46,21 @@ struct ion_handle_data {
...
+}
+
+typedef int (*rem_ion_open_fn)();
+rem_ion_open_fn ion_open_fn = NULL;
+
+typedef int (*rem_ion_alloc_fd_fn)(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags, int* map_fd);
+rem_ion_alloc_fd_fn ion_alloc_fd_fn = NULL;
+
+// ION IOCTL approach
+int ion_alloc(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags) {
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/host_malloc.cpp]: Revert this too.

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 8]
```diff
@@ -137,19 +172,28 @@ void *halide_hexagon_host_malloc(size_t size) {
...
 
-    ion_user_handle_t handle = ion_alloc(ion_fd, size, alignment, 1 << heap_id, ion_flags);
-    if (handle < 0) {
-        __android_log_print(ANDROID_LOG_ERROR, "halide", "ion_alloc(%d, %d, %d, %d, %d) failed",
+    int buf_fd = 0;
+    ion_user_handle_t handle = 0;
+    if (use_libion) {
+        if (ion_alloc_fd_fn(ion_fd, size, alignment, 1 << heap_id, ion_flags, &buf_fd)) {
+            __android_log_print(ANDROID_LOG_ERROR, "halide", "ion_alloc_fd(%d, %d, %d, %d, %d) failed",
                             ion_fd, size, alignment, 1 << heap_id, ion_flags);
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/host_malloc.cpp]: Indent

File: src/runtime/hexagon_remote/host_malloc.cpp [hunk 9]
```diff
@@ -137,19 +172,28 @@ void *halide_hexagon_host_malloc(size_t size) {
...
-        __android_log_print(ANDROID_LOG_ERROR, "halide", "ion_map(%d, %d) failed", ion_fd, handle);
-        ion_free(ion_fd, handle);
-        return NULL;
+            return NULL;
+        }
+    } else { // !use_libion
+        handle = ion_alloc(ion_fd, size, alignment, 1 << heap_id, ion_flags);
+        if (handle < 0) {
+            __android_log_print(ANDROID_LOG_ERROR, "halide", "ion_alloc(%d, %d, %d, %d, %d) failed",
+                            ion_fd, size, alignment, 1 << heap_id, ion_flags);
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/host_malloc.cpp]: Indent

**Comment by ronlieb** [comment]
i am not sure how concerned folks are about the tool chains i used to build the runtimes.
So before this gets merged, anyone want these rebuilt with the exact same recipe Google uses internally ?

**Comment by ronlieb** [comment]
ping

**Comment by dsharletg** [comment]
How can I test this? I'm testing it on device, I'm not seeing any indication that it's loading libion.so, despite that library being present on the device.

**Comment by ronlieb** [comment]
it might be worthwhile for me to add two messages in the appropriate places, if you wish i can do so....

__android_log_print(ANDROID_LOG_ERROR, "halide", "Using ion ioctl");

__android_log_print(ANDROID_LOG_ERROR, "halide", "Using libion.so");

**Comment by dsharletg** [comment]
Looks like it is working. I'd rather not commit logging in non-error code paths, but I think it might be worth it in this case.

**Comment by ronlieb** [comment]
shall we merge this one ?

**Comment by pranavb-ca** [comment]
@ronlieb  - Let's wait till all the buildbots are up.

**Comment by ronlieb** [comment]
i did not realize the buildbots were still down, yes, lets wait

**Comment by steven-johnson** [comment]
I just updated to master now that the buildbots are finally up again.

**Comment by ronlieb** [comment]
found a way to get the ioctl to work on the newer kernel,  will push an updated patch soon

**Comment by ronlieb** [comment]
Closing this in favor of https://github.com/halide/Halide/pull/3138

**Pull request closed without merging**